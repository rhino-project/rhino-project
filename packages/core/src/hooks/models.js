import { isObject } from 'lodash';
import { useMemo } from 'react';
import { getModel } from 'rhino/utils/models';

/**
 * @typedef {import('../utils/models.js').Model} Model
 */

/**
 * Memoize a model by name or with an existing model object
 *
 * @param {string | Model} model - The model name or model to memoize
 * @returns {Model} Memoized model object
 *
 * @example
 *    const model = useModel('blog_post')
 *    const model = useModel(getModel('blog_post'))
 */
export const useModel = (model) =>
  useMemo(() => (isObject(model) ? model : getModel(model)), [model]);
