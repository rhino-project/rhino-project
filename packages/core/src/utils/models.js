import { filter, find, toPath, set } from 'lodash';
import modelLoader from 'rhino/models';

/**
 * @typedef {object} Model
 * @property {string} model
 * @property {string} name
 */

export const getModel = (model) => modelLoader.api.components.schemas[model];
export const getParentModel = (model) => getModel(model.ownedBy);
export const getModelAncestors = (model) => {
  const result = [];
  model = getParentModel(model);
  while (model != null) {
    result.push(model);
    model = getParentModel(model);
  }
  return result;
};
export const getBaseOwnerFilters = (model, baseOwnerId) => {
  const filters = {};
  const pathToRoot = getModelAncestors(model);
  const rootModel = pathToRoot[pathToRoot.length - 1];
  // if the root model is globally owned, we don't want to add any filters. these models have 'global' as parent.
  // only the base owner will have null as parent.
  if (baseOwnerId && rootModel && rootModel.ownedBy == null) {
    const modelNames = pathToRoot.map((el) => el.model);
    set(filters, ['filter', ...modelNames], baseOwnerId);
  }
  return filters;
};

export const getCreatableAttributes = (model) =>
  filter(model.properties, (a) => a.creatable);
export const getUpdatableAttributes = (model) =>
  filter(model.properties, (a) => a.updatable);

export const isReference = (attribute) => attribute.type === 'reference';
export const getReferenceAttributes = (model) =>
  filter(model.properties, (a) => isReference(a));

export const isIdentifier = (attribute) => attribute.type === 'identifier';
export const getIdentifierAttribute = (model) =>
  find(model.properties, (a) => isIdentifier(a));

export const getModuleInfo = (module) =>
  modelLoader.api.info['x-rhino']?.modules?.[module];

export const authOwnerProperty = () => getModuleInfo('rhino')?.authOwner;
export const authOwnerModel = () => getModel(authOwnerProperty());

export const baseOwnerProperty = () => getModuleInfo('rhino')?.baseOwner;
export const baseOwnerModel = () => getModel(baseOwnerProperty());
export const isBaseOwned = (model) => model.ownedBy === baseOwnerProperty();
export const getBaseOwnedModels = () =>
  filter(modelLoader.api.components.schemas, isBaseOwned);

export const getOwnedModels = (model) =>
  filter(modelLoader.api.components.schemas, (m) => m.ownedBy === model.model);

export const getModelFromRef = (attribute) => {
  const ref =
    // Direct ref
    attribute['$ref'] ||
    // anyOf ref, probably because its nullable
    attribute?.anyOf?.[0]?.['$ref'] ||
    // Any array of refs
    attribute?.items?.['$ref'] ||
    // Any array of anyOf refs
    attribute?.items?.anyOf?.[0]?.['$ref'];

  if (!ref) return null;

  const parts = ref.split('/');
  const refName = parts[parts.length - 1];

  const model = modelLoader.api.components.schemas[refName];

  console.assert(
    model,
    `Reference for attribute ${attribute.name} was not found`,
    attribute
  );

  return model;
};

export const getModelAndAttributeFromPath = (model, path) => {
  const split = toPath(path);
  const length = split.length;
  let last_model = model;
  let index = 0;
  let attribute = null;
  let operator = null;
  let plainPath = [];

  while (model && index < length) {
    const [attributeName, operatorName] = split[index++].split('::');

    // A number means its an array indicator
    if (!isNaN(parseInt(attributeName))) {
      continue;
    }
    plainPath.push(attributeName);

    // FIXME Gross hack
    if (attributeName === 'display_name') {
      return [
        last_model,
        {
          model: model.model,
          name: model.model,
          // The last attribute in the path should be the parent model
          readableName: attribute?.readableName || 'Name',
          type: 'string'
        }
      ];
    } else {
      attribute = model.properties[attributeName];
      operator = operatorName;
    }

    last_model = model;
    model = getModelFromRef(attribute);
  }

  return index && index === length
    ? [last_model, attribute, operator, plainPath.join('.')]
    : [undefined, undefined];
};

export const getAttributeFromPath = (model, path) =>
  getModelAndAttributeFromPath(model, path)[1];

export const oauthProviders = () => getModuleInfo('rhino')?.oauth || [];

export const hasModule = (moduleName) =>
  getModuleInfo(moduleName) !== undefined;

export const hasOrganizationsModule = () => hasModule('rhino_organizations');
export const hasNotificationsModule = () => hasModule('rhino_notifications');
export const hasSubscriptionsModule = () => hasModule('rhino_subscriptions');
