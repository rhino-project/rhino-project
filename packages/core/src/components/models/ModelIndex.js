import PropTypes from 'prop-types';

import { useGlobalComponent, useGlobalOverrides } from 'rhino/hooks/overrides';

import ModelIndexHeader from 'rhino/components/models/ModelIndexHeader';
import ModelWrapper from 'rhino/components/models/ModelWrapper';
import { ModelIndexTableBase } from './ModelIndexTable';
import ModelIndexActions from './ModelIndexActions';
import ModelIndexSimple from './ModelIndexSimple';

const defaultComponents = {
  ModelIndexHeader,
  ModelIndexActions,
  ModelIndexTable: ModelIndexTableBase
};

export const ModelIndexBase = ({
  overrides,
  baseFilter,
  wrapper,
  ...props
}) => {
  const {
    ModelIndexHeader,
    ModelIndexActions,
    ModelIndexTable
  } = useGlobalOverrides(defaultComponents, overrides, props);
  const { model } = props;

  if (baseFilter)
    console.warn(
      'baseFilter is deprecated. Use filter/limit/offset/order/search instead'
    );

  // Legacy baseFilter support - must be after the new props to avoid being overwritten by undefined
  return (
    <ModelIndexSimple {...props} {...baseFilter}>
      <ModelWrapper model={model} wrapper={wrapper} baseClassName="index">
        {/* FIXME: Stop passing down props */}
        <ModelIndexHeader {...props} />
        <hr />
        <ModelIndexActions {...props} />
        <ModelIndexTable {...props} />
      </ModelWrapper>
    </ModelIndexSimple>
  );
};

ModelIndexBase.propTypes = {
  baseFilter: PropTypes.object,
  model: PropTypes.object.isRequired,
  overrides: PropTypes.object,
  parent: PropTypes.object
};

const ModelIndex = (props) => useGlobalComponent(ModelIndexBase, props);

export default ModelIndex;
