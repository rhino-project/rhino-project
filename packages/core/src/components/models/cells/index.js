export * from './ModelCellArray';
export * from './ModelCellArrayReference';
export * from './ModelCellAttachment';
export * from './ModelCellAttachmentDownload';
export * from './ModelCellAttachmentImage';
export * from './ModelCellAttachments';
export * from './ModelCellBoolean';
export * from './ModelCellBooleanIcon';
export * from './ModelCellCountry';
export * from './ModelCellCurrency';
export * from './ModelCellDate';
export * from './ModelCellDateTime';
export * from './ModelCellDateTimeDistance';
export * from './ModelCellEnum';
export * from './ModelCellFloat';
export * from './ModelCellInteger';
export * from './ModelCellLinkEmail';
export * from './ModelCellLinkTelephone';
export * from './ModelCellReference';
export * from './ModelCellString';
export * from './ModelCellTime';
export * from './ModelEditableCellBoolean';
export * from './ModelEditableCellReference';
export * from './ModellCellIdentifier';
