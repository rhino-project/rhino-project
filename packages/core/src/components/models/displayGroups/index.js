export * from './ModelDisplayGroupArray';
export * from './ModelDisplayGroupArrayReference';
export * from './ModelDisplayGroupAttachment';
export * from './ModelDisplayGroupAttachmentImage';
export * from './ModelDisplayGroupAttachments';
export * from './ModelDisplayGroupBoolean';
export * from './ModelDisplayGroupCurrency';
export * from './ModelDisplayGroupDate';
export * from './ModelDisplayGroupDateTime';
export * from './ModelDisplayGroupEnum';
export * from './ModelDisplayGroupFloat';
export * from './ModelDisplayGroupInteger';
export * from './ModelDisplayGroupReference';
export * from './ModelDisplayGroupString';
export * from './ModelDisplayGroupText';
export * from './ModelDisplayGroupTime';
