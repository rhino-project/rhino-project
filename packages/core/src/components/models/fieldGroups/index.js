export * from './ModelFieldGroupArrayInteger';
export * from './ModelFieldGroupArrayString';
export * from './ModelFieldGroupBoolean';
export * from './ModelFieldGroupCountry';
export * from './ModelFieldGroupCurrency';
export * from './ModelFieldGroupDate';
export * from './ModelFieldGroupDateTime';
export * from './ModelFieldGroupEnum';
export * from './ModelFieldGroupFile';
export * from './ModelFieldGroupFloat';
export * from './ModelFieldGroupInteger';
export * from './ModelFieldGroupIntegerSelect';
export * from './ModelFieldGroupJoinSimple';
export * from './ModelFieldGroupNested';
export * from './ModelFieldGroupOwnerReference';
export * from './ModelFieldGroupPhone';
export * from './ModelFieldGroupReference';
export * from './ModelFieldGroupString';
export * from './ModelFieldGroupText';
export * from './ModelFieldGroupTime';
export * from './ModelFieldGroupYear';
