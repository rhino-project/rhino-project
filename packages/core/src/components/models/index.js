export * from './ModelCell';
export * from './ModelCreate';
export * from './ModelCreateActions';
export * from './ModelCreateForm';
export * from './ModelCreateHeader';
export * from './ModelCreateModal';
export * from './ModelCreateModalActions';
export * from './ModelCreateProvider';
export * from './ModelCreateSimple';
export * from './ModelDisplayGroup';
export * from './ModelDisplayGroupFloating';
export * from './ModelDisplayGroupHorizontal';
export * from './ModelDisplayLabel';
export * from './ModelEdit';
export * from './ModelEditActions';
export * from './ModelEditForm';
export * from './ModelEditHeader';
export * from './ModelEditModal';
export * from './ModelEditModalActions';
export * from './ModelEditProvider';
export * from './ModelEditSimple';
export * from './ModelFieldGroup';
export * from './ModelFieldGroupFloating';
export * from './ModelFieldGroupHorizontal';
export * from './ModelFieldLabel';
export * from './ModelFilter';
export * from './ModelFilterGroup';
export * from './ModelFilterLabel';
export * from './ModelFilters';
export * from './ModelFiltersProvider';
export * from './ModelFiltersSimple';
export * from './ModelFooter';
export * from './ModelHeader';
export * from './ModelIndex';
export * from './ModelIndexActions';
export * from './ModelIndexCardGrid';
export * from './ModelIndexHeader';
export * from './ModelIndexProvider';
export * from './ModelIndexSimple';
export * from './ModelIndexTable';
export * from './ModelPager';
export * from './ModelProvider';
export * from './ModelSearch';
export * from './ModelSection';
export * from './ModelShow';
export * from './ModelShowActions';
export * from './ModelShowDescription';
export * from './ModelShowHeader';
export * from './ModelShowProvider';
export * from './ModelShowRelated';
export * from './ModelShowSimple';
