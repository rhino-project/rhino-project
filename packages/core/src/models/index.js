import { each } from 'lodash';

import staticModels from 'models/static';

const MODEL_PATH = '/api/info/openapi';

const hoistRhino = (data) => {
  each(Object.keys(data.components.schemas), (schema) => {
    const model_uplifted = {
      ...data.components.schemas[schema]['x-rhino-model']
    };

    data.components.schemas[schema] = {
      ...data.components.schemas[schema],
      ...model_uplifted
    };

    // OpenAPI does required on the model level, we need it on the
    // attribute level
    const required = data.components.schemas[schema].required;

    each(
      Object.keys(data.components.schemas[schema].properties),
      (property) => {
        const attribute_uplifted = {
          ...data.components.schemas[schema].properties[property][
            'x-rhino-attribute'
          ]
        };

        if (required?.includes(property)) {
          attribute_uplifted['x-rhino-required'] = true;
        }

        data.components.schemas[schema].properties[property] = {
          ...data.components.schemas[schema].properties[property],
          ...attribute_uplifted
        };
      }
    );
  });

  return data;
};

class ModelLoader {
  async loadModels(forceStatic = false) {
    if (forceStatic || import.meta.env.PROD) {
      this.api = hoistRhino(staticModels);
    } else {
      const { networkApiCall } = await import('rhino/lib/networking');

      return networkApiCall(MODEL_PATH).then(
        (response) => (this.api = hoistRhino(response.data))
      );
    }
  }
}

export default new ModelLoader();
