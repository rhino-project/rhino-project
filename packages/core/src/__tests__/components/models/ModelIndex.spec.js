import { render } from '@testing-library/react';
import { sharedModelTests } from './sharedModelTests';
import { ModelIndex } from '../../../components/models/ModelIndex';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MemoryRouter } from 'react-router-dom';

describe('ModelIndex', () => {
  const Foo = () => <div>Foo</div>;
  const Bar = () => <div>Bar</div>;
  const Baz = () => <div>Baz</div>;

  const wrapper = ({ children }) => {
    const queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false
        }
      }
    });

    return (
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>{children}</MemoryRouter>
      </QueryClientProvider>
    );
  };

  sharedModelTests(ModelIndex);

  it(`should allow local overrides`, async () => {
    const { asFragment } = render(
      <ModelIndex
        overrides={{
          ModelIndexHeader: Foo,
          ModelIndexActions: Bar,
          ModelIndexTable: Baz
        }}
        model={{ model: 'user', properties: { name: {} } }}
        path="name"
        fallback={false}
        queryOptions={{ enabled: false }}
      />,
      { wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
