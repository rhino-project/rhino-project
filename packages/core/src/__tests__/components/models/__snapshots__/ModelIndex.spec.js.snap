// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelIndex > should allow local overrides 1`] = `
<DocumentFragment>
  <div
    class="model-index model-index-user"
  >
    <div>
      Foo
    </div>
    <hr />
    <div>
      Bar
    </div>
    <div>
      Baz
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModelIndex > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelIndex > should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
